File: erlang/sum-of-multiples/sum_of_multiples.erl
-module(sum_of_multiples).
-export([sum_of_multiples/1, sum_of_multiples/2]).

%% When called with only a number without multipliers specified
%% we default to using 3 and 5 as our multipliers.
sum_of_multiples(Number) when is_number(Number) ->
  sum_of_multiples([5], Number, total_for_multiplier(3, Number - 1, 0)).

%% When an empty list of multipliers is supplied we return 0.
sum_of_multiples([], _) ->
  0;
sum_of_multiples([Multiplier|Tail], Number) when is_number(Number) ->
  sum_of_multiples(Tail, Number, total_for_multiplier(Multiplier, Number - 1, 0)).

sum_of_multiples([], _, Sum) ->
  Sum;
sum_of_multiples([Multiplier|Tail], Number, Sum) ->
  sum_of_multiples(Tail, Number, Sum + total_for_multiplier(Multiplier, Number - 1, 0)).

total_for_multiplier(Multiplier, Number, Total) when Number < Multiplier ->
  Total;
total_for_multiplier(Multiplier, Number, Total) when Number rem Multiplier =:= 0 ->
  total_for_multiplier(Multiplier, Number - 1, Total + Number);
total_for_multiplier(Multiplier, Number, Total) ->
  total_for_multiplier(Multiplier, Number - 1, Total).

File: haskell/sum-of-multiples/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

sumOfMultiples :: [Int] -> Int -> Int
sumOfMultiples [] _ = 0
sumOfMultiples (multiplier:multipliers) number =
  calculateMultiples multipliers number (totalForMultiplier multiplier (number - 1) 0)

calculateMultiples :: [Int] -> Int -> Int -> Int
calculateMultiples [] _ total = total
calculateMultiples (multiplier:multipliers) number total =
  calculateMultiples multipliers number (total + (totalForMultiplier multiplier (number - 1) 0))

totalForMultiplier :: Int -> Int -> Int -> Int
totalForMultiplier multiplier number total
  | number < multiplier = total
  | mod number multiplier == 0 = totalForMultiplier multiplier (number - 1) (total + number)
  | otherwise = totalForMultiplier multiplier (number - 1) total